python/revng/cli/_commands/idb_converter.py

```diff
@@ -0,0 +1,655 @@
+        self.idb: idb.fileformat.IDB = input_idb
+        self.api = idb.IDAPython(self.idb)
+
+        self.arch: m.Architecture = self._get_arch()
+        self.is64bit: bool = self._is_64_bit()
+        self.segments: List[m.Segment] = []
+        self.revng_types_by_id: Dict[int, RevngTypes] = {}
+        self.idb_types_to_revng_types: Dict[int, m.QualifiedType] = {}
+        self.functions: Set[m.Function] = set()
+        self.dynamic_functions: List[m.DynamicFunction] = []
+        self.imported_libraries: List[str] = []
+        self.base_addr = base_addr
+
+        self._structs_to_fixup: Set[Tuple[m.StructType, idb.typeinf.TInfo]] = set()
+        self._unions_to_fixup: Set[Tuple[m.UnionType, idb.typeinf.TInfo]] = set()
+
+        self._import_types()
+        self._fixup_structs()
+        self._fixup_unions()
+        self._import_functions()
+        self._collect_imports()
+        self._fixup_void_cabifunctiontype_arguments()
+
+    def _import_types(self):
+        """Imports initial types from the IDB. The types will be incomplete and need to be fixed"""
+        til = self.idb.til
+        for type_definition in til.types.defs:
+            assert isinstance(type_definition, idb.typeinf.TILTypeInfo)
+            the_type = type_definition.type
+            assert isinstance(the_type, idb.typeinf.TInfo)
+            self._convert_idb_type_to_revng_type(the_type, ordinal=type_definition.ordinal)
+
+    def _import_functions(self):
+        metaaddr_type = revng_arch_to_metaaddr_code_type[self.arch]
+
+        for function_start_addr in self.api.idautils.Functions():
+            function = idb.analysis.Function(self.idb, function_start_addr)
+            function_name = function.get_name()
+
+            # NOTE: python-idb raises an exception when deserializing type
+            # info from some DBs (it tries to read a byte from an empty buffer).
+            try:
+                idb_function_type = function.get_signature()
+            except Exception as exception:
+                log_error("Unable to parse a function type.")
+                log_error(str(exception))
+                idb_function_type = None
+
+            if idb_function_type is not None:
+                qualified_revng_function_type = self._convert_idb_type_to_revng_type(
+                    idb_function_type
+                )
+                revng_function_type = self.unwrap_qualified(qualified_revng_function_type)
+            else:
+                abi = revng_arch_to_abiname[self.arch]
+
+                qualified_return_type = self._get_primitive_type(m.PrimitiveTypeKind.Void, 0)
+                revng_function_type = m.CABIFunctionType(
+                    OriginalName=function_name,
+                    ABI=abi,
+                    ReturnType=qualified_return_type,
+                    Arguments=[],
+                )
+                self.revng_types_by_id[revng_function_type.ID] = revng_function_type
```

Comment:
What's going on here? When does this happen?
Let's at least emit a warning.

Reply:
Yeah, warning is ok for sure. But I think that we should avoid creating prototype here, since IDA was not able to find the signature. For example, the case ```always-dead-return-value-revng-qa.compiled``` is a small example for this -- all 3 functions have no prototype info, and when we try ```import binary``` it looks as:
```
Functions:
  - Entry:           "0x401000:Code_x86_64"
    CustomName:      "prefix__start"
    OriginalName:    _start
    Type:            Regular
  - Entry:           "0x401006:Code_x86_64"
    CustomName:      "func"
    OriginalName:    func
    Type:            Regular
  - Entry:           "0x401013:Code_x86_64"
    CustomName:      "callee"
    OriginalName:    callee
    Type:            Regular
```
And I think that we should avoid production of the artificial type in the case of IDB as well.

Reply:
Having no prototype for a function is fine. It's a valid model.
No artificial type + warning makes sense I think.

Reply:
Yeah, on the same page.